ambry-server/src/main/java/com/github/ambry/server/StatsManager.java

```diff
@@ -435,53 +478,108 @@ long getPermanentDeleteTombstoneSize() {
+        } finally {
           if (accountStatsMySqlStore != null) {
-            accountStatsMySqlStore.storeAccountStats(statsWrapper);
+            accountStatsMySqlStore.closeConnection();
           }
-          publish(statsWrapper);
-          logger.info("Local stats snapshot published to {}", statsOutputFile.getAbsolutePath());
         }
-      } catch (Exception | Error e) {
-        metrics.statsAggregationFailureCount.inc();
-        logger.error("Exception while aggregating stats for local report. Stats output file path - {}",
-            statsOutputFile.getAbsolutePath(), e);
-      } finally {
-        if (accountStatsMySqlStore != null) {
+      }
+    }
+
+    void cancel() {
+      cancelled = true;
+    }
+  }
+
+  /**
+   * Runnable class that collects, publishes partition class stats to mysql database.
+   */
+  class PartitionClassStatsPublisher implements Runnable {
+    private volatile boolean cancelled = false;
+    private final AccountStatsMySqlStore accountStatsMySqlStore;
+
+    PartitionClassStatsPublisher(AccountStatsMySqlStore accountStatsMySqlStore) {
+      this.accountStatsMySqlStore = accountStatsMySqlStore;
+    }
+
+    @Override
+    public void run() {
+      synchronized (accountStatsMySqlStore) {
+        logger.info("Aggregating partition class stats for local report");
+        try {
+          long totalFetchAndAggregateStartTimeMs = time.milliseconds();
+          StatsSnapshot aggregatedSnapshot = new StatsSnapshot(0L, new HashMap<>());
+          List<PartitionId> unreachablePartitions = new ArrayList<>();
+          Iterator<PartitionId> iterator = (new HashSet<>(partitionToReplicaMap.keySet())).iterator();
+          while (!cancelled && iterator.hasNext()) {
+            PartitionId partitionId = iterator.next();
+            logger.info("Aggregating partition class stats for local report started for store {}", partitionId);
+            collectAndAggregatePartitionClassStats(aggregatedSnapshot, partitionId, unreachablePartitions);
+          }
+          aggregatedSnapshot.updateValue();
+          List<String> unreachableStores = examineUnreachablePartitions(unreachablePartitions);
+          if (!cancelled) {
+            metrics.totalFetchAndAggregateTimeMs.update(time.milliseconds() - totalFetchAndAggregateStartTimeMs);
+            StatsHeader statsHeader =
+                new StatsHeader(StatsHeader.StatsDescription.STORED_DATA_SIZE, time.milliseconds(),
+                    partitionToReplicaMap.keySet().size(),
+                    partitionToReplicaMap.keySet().size() - unreachableStores.size(), unreachableStores);
+            StatsWrapper statsWrapper = new StatsWrapper(statsHeader, aggregatedSnapshot);
+            accountStatsMySqlStore.storePartitionClassStats(statsWrapper);
+            logger.info("Local partition class stats snapshot published to {}", statsOutputFile.getAbsolutePath());
+          }
+        } catch (Exception | Error e) {
```

Comment:
Throwable is equivalent.  But if we get an Error like OOM, do we really want to eat it?

Reply:
Will ignore Error.